BigFrames Quickstart (#7139)
dbt-labs/docs.getdbt.com: website/docs/guides/dbt-python-bigframes.md

2. In the file, copy this sql into it. 

```sql
    select 
    1 as foo,
    2 as bar
```
3. Now create a new model file in the models directory, named `my_first_python_model.py`. 
4. Configure the Bigframes submission method by either:  

(a) Project level configuration via dbt_project.yml

```python
models:
  my_dbt_project:
    submission_method: bigframes
    python_models:
      +materialized: view
```
or 

(b) The Python code via dbt.config

```python
def model(dbt, session):
   dbt.config(submission_method="bigframes")
   bdf = dbt.ref("my_sql_model") 
   return bdf
```

3. Now in that file, add in this code:

```python
def model(dbt, session):
   dbt.config(submission_method="bigframes")
   bdf = dbt.ref("my_sql_model") #loading from prev step
   return bdf
```

3. Create a new python model file named `my_second_python_model.py` 

```python
def model(dbt, session):  
   data = {"foo": [1, 2], "bar": [3, 4]}
   return bpd.DataFrame(data=data)
```

4. Run `dbt run` 

5. *(Optional)* **You can also go to the Colab Enterprise Executions tab and GCS bucket from the GCP console to view the codes and logs (including previous executions).**

> 6. Congrats! You just created your first two python models to run on BigFrames! 

Reviewer: ```suggestion
6. Congrats! You just created your first two python models to run on BigFrames! 

</div>
```

---

> title: "Quickstart for using BigQuery Dataframes with dbt python models"

Reviewer: ```suggestion
title: "Using BigQuery Dataframes with dbt python models"
```

id: dbt-python-bigframes

> description: "Use this guide to help you setup dbt with BigQuery Dataframes (BigFrames)."

Reviewer: ```suggestion
description: "Use this guide to help you set up dbt with BigQuery Dataframes (BigFrames)."
```

> sidebar_label: "Quickstart for BigQuery Dataframes and dbt python"

Reviewer: ```suggestion
sidebar_label: "BigQuery Dataframes and dbt python"
```

# time_to_complete: '30 minutes' comment out until test
icon: 'guides'
hide_table_of_contents: true
tags: ['BigQuery', 'Google', 'GCP', 'Bigframes','Quickstart']
keywords: ['BigQuery','dbt', 'Google']
level: 'Intermediate'
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

## Introduction

> This guide will help you set up dbt to be used with BigQuery Dataframes (Bigframes).

Reviewer: ```suggestion
In this guide, you'll learn how to set up dbt so you can use it with BigQuery Dataframes (Bigframes).
```

> In addition to the existing dataproc/pyspark based submission methods for executing python models, you can now use the bigframes submission method to execute pandas and scikit- learn code at scale on the BigQuery SQL engine.

Reviewer: ```suggestion
In addition to the existing dataproc/pyspark based submission methods for executing python models, you can now use the BigFrames submission method to execute pandas and sciki, which help you learn code at scale on the BigQuery SQL engine.
```

> BigQuery Dataframes is an open source python package that transpiles pandas and sci-kit learn code to scalable BigQuery SQL. The dbt-bigquery adapter relies on the BigQuery Studio Notebook Executor Service to run the python client side code.

Reviewer: ```suggestion
BigQuery Dataframes is an open source python package that transpiles pandas and scikit-learn code to scalable BigQuery SQL. The dbt-bigquery adapter relies on the BigQuery Studio Notebook Executor Service to run the python client side code.
```

> ### What you'll use during the lab

Reviewer: ```suggestion
### Prerequisites
```

- A [Google Cloud account](https://cloud.google.com/free) 
- A [dbt Cloud account](https://www.getdbt.com/signup/) 

### What you'll learn

- How to build scalable data transformation pipelines using dbt and Google Cloud, with SQL and Python.
- How to leverage Bigframes from dbt for scalable BigQuery SQL.

> ### What you need to know

Reviewer: Move this content up to "prerequisites"

- Basic to intermediate SQL and python.
- Basic understanding of dbt fundamentals. We recommend the [dbt Fundamentals course](https://learn.getdbt.com) if you're interested.

### What you'll build

> This guide contains two sections:

Reviewer: ```suggestion
Here's what you'll build in two parts:
```

---
title: "Using BigQuery Dataframes with dbt python models"
id: dbt-python-bigframes
description: "Use this guide to help you set up dbt with BigQuery Dataframes (BigFrames)."
sidebar_label: "BigQuery Dataframes and dbt python"
# time_to_complete: '30 minutes' comment out until test
icon: 'guides'
hide_table_of_contents: true
tags: ['BigQuery', 'Google', 'GCP', 'BigFrames','Quickstart']
keywords: ['BigQuery','dbt', 'Google']
level: 'Intermediate'
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

## Introduction

In this guide, you'll learn how to set up dbt so you can use it with BigQuery Dataframes (BigFrames).

> In addition to the existing dataproc/pyspark based submission methods for executing python models, you can now use the BigFrames submission method to execute pandas and sciki, which help you learn code at scale on the BigQuery SQL engine.

Reviewer: ```suggestion
In addition to the existing dataproc/pyspark based submission methods for executing python models, you can now use the BigFrames submission method to execute pandas and scikit, which help you learn code at scale on the BigQuery SQL engine.
```

---
title: "Quickstart for using BigQuery Dataframes with dbt python models"
id: dbt-python-bigframes

> description: "Use this guide to help you setup dbt Core with BigQuery Dataframes (BigFrames)."

Reviewer 2: ```suggestion
description: "Use this guide to help you setup dbt with BigQuery Dataframes (BigFrames)."
```

sidebar_label: "Quickstart for BigQuery Dataframes and dbt python"
# time_to_complete: '30 minutes' comment out until test
icon: 'guides'
hide_table_of_contents: true
tags: ['BigQuery', 'Google', 'GCP', 'Bigframes','Quickstart']

> keywords: ['BigQuery','dbt Core', 'Google']

Reviewer 2: ```suggestion
keywords: ['BigQuery','dbt', 'Google']
```

level: 'Intermediate'
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

## Introduction

> This guide will help you set up dbt Core to be used with BigQuery Dataframes (Bigframes).

Reviewer 2: ```suggestion
This guide will help you set up dbt to be used with BigQuery Dataframes (Bigframes).
```

3. *(Optional)* **Create a test BigQuery Dataset**

Create a new BigQuery Dataset if you don't already have one.

```python
#Create BQ dataset 
bq mk --location=${REGION} echo "${GOOGLE_CLOUD_PROJECT}" | tr '-' '_'_dataset
```

4. *(Optional)* **Create a GCS bucket to stage the python code**

If you wish to store the python compiled code on a GCS bucket, create a new one.

```python
#Create GCS bucket
gcloud storage buckets create gs://${GOOGLE_CLOUD_PROJECT}-bucket --location=${REGION}
#Grant Storage Admin over the bucket to your SA 
xgcloud storage buckets add-iam-policy-binding gs://${GOOGLE_CLOUD_PROJECT}-bucket --member=serviceAccount:dbt-bigframes-sa@${GOOGLE_CLOUD_PROJECT}.iam.gserviceaccount.com --role=roles/storage.admin
```

5. **Create a GCS bucket to hold the logs**
```python
#Create GCS bucket
gcloud storage buckets create gs://${GOOGLE_CLOUD_PROJECT}-bucket-logs --location=${REGION}
#Grant Storage Admin over the bucket to your SA 
gcloud storage buckets add-iam-policy-binding gs://${GOOGLE_CLOUD_PROJECT}-bucket-logs --member=serviceAccount:dbt-bigframes-sa@${GOOGLE_CLOUD_PROJECT}.iam.gserviceaccount.com --role=roles/storage.admin
```

> ## Client Side Configuration 

Reviewer 2: Please remove the section about dbt core setup. The requirements from the beginning is already based on dbt Cloud and they should have this setup before getting started. You can also point them to the BQ + dbt quickstart as an alternative.

- Leave timeout to the default value of 300 
- Select location e.g. us-central1

4. **Edit dbt profile.yml under $HOME/.dbt

For example (`service account` credential)

```yaml
my_dbt_project_sa:
  outputs:
    dev:
      dataproc_region: us-central1
      dataset: jialuo_test_us
      gcs_bucket: jialuo_bf_dbt_integration_dev1
      job_execution_timeout_seconds: 300
      job_retries: 1
      keyfile: /usr/local/google/home/jialuo/dbt_t2/secrets/sa_key.json
      location: US
      method: service-account
      priority: interactive
      project: bigframes-dbt-integration-dev
      threads: 1
      type: bigquery
  target: dev
```

OR ('Oauth' credential)

```yaml
my_dbt_project:
  outputs:
    dev:
      dataproc_region: us-central1
      dataset: jialuo_test_us
      gcs_bucket: jialuo_bf_dbt_dev2
      job_execution_timeout_seconds: 300
      job_retries: 1
      location: US
      method: oauth
      priority: interactive
      project: bigframes-dev-perf
      threads: 1
      type: bigquery
  target: dev
```
Please note, the bigframes submission method does **NOT** introduce any new configuration in profile.yml under $HOME/.dbt.

> 5. Set Bigframes submission method

Reviewer 2: I would move this submission piece into the next section so you can explicitly tell them this is how to execute a python model.

All that you need to say is there's two ways to configure this - one the project.yml level or in the model config.

# files using the `{{ config(...) }}` macro.
models:
  my_dbt_project:
    submission_method: bigframes
    # Config indicated by + and applies to all files under models/example/
    example:
      +materialized: view
```

(b) The Python code via dbt.config, which you will be creating later and can overwrite (a)

For example

```python
def model(dbt, session):
   # Will overwrite the submission method setup in dbt_project.yml
   dbt.config(submission_method="bigframes")

   bdf = dbt.ref("my_sql_model") #loading from prev step
   <process bigquery dataframe >
   return bdf
```

## Execute a python model

1. **Add a new .py file on your dbt project**

Now, you can add python models to your dbt project models folder.

For example

```python
def model(dbt, session):
   # Will overwrite the submission method setup in dbt_project.yml
   dbt.config(submission_method="bigframes")

   bdf = dbt.ref("my_sql_model") #loading from prev step
   <process bigquery dataframe >
   return bdf
```

```python
# Already set bigframes submission method in dbt_project.yml
def model(dbt, session):  
   data = {"foo": [1, 2], "bar": [3, 4]}
   return bpd.DataFrame(data=data)
```

> 2. **Execute the dbt pipeline**

Reviewer 2: To keep this short and sweet - let's just have it execute the example dbt model specifically. This isn't a dbt 101

22:35:52  BigQuery adapter: The colab notebook runtime outputs from GCS: 

```Some outputs of the Colab notebook from bigframes, which is stored in GCS```

22:35:52  1 of 3 OK created python table model jialuo_test_us.my_bigframes_model ......... [OK in 228.85s]
22:35:52  2 of 3 START sql table model jialuo_test_us.my_first_dbt_model ................. [RUN]
22:35:55  2 of 3 OK created sql table model jialuo_test_us.my_first_dbt_model ............ [CREATE TABLE (2.0 rows, 0 processed) in 3.25s]
22:35:55  3 of 3 START sql view model jialuo_test_us.my_second_dbt_model ................. [RUN]
22:35:56  3 of 3 OK created sql view model jialuo_test_us.my_second_dbt_model ............ [CREATE VIEW (0 processed) in 0.96s]
22:35:56  
22:35:56  Finished running 2 table models, 1 view model in 0 hours 3 minutes and 55.30 seconds (235.30s).
22:35:56  
22:35:56  Completed successfully
22:35:56  
22:35:56  Done. PASS=3 WARN=0 ERROR=0 SKIP=0 TOTAL=3
```

OR, just execute the selected model
```
> dbt run --select [your_bigframes_model]

...
```

3. **You can observe the python execution in the dbt logs**
For example
```
data:

>     text/html: ['Query job eaa77dfe-1e17-492e-815d-4728d65a07e3 is DONE. 0 Bytes processed. <a target="_blank" href="https://console.cloud.google.com/bigquery?project=bigframes-dev-perf&j=bq:US:eaa77dfe-1e17-492e-815d-4728d65a07e3&page=queryresults">Open Job</a>']

Reviewer 2: should we be removing any specific account information from the logs?

metadata:
output_type:
    display_data
```

4. *(Optional)* **You can also go to the Colab Enterprise Executions tab and GCS bucket from the GCP console to view the codes and logs (including previous executions).**

The 'Runtime Templates' tab:

<Lightbox src="/img/guides/gcp-guides/gcp-bigframes-runtime.png" title="Runtime Templates"/>

Note:

If no `default` runtime template is available, it will automatically create one for the user and mark it `default` for next time usage.
To get the permissions that you need to create a runtime template in Colab Enterprise, ask your administrator to grant you the Colab Enterprise Admin (roles/aiplatform.colabEnterpriseAdmin) IAM role on the project ([runtime template link](https://cloud.google.com/colab/docs/create-runtime-template#:~:text=By%20creating%20a%20runtime%20template,characteristics%20based%20on%20your%20needs.)).

The 'Executions' tab:
<Lightbox src="/img/guides/gcp-guides/gcp-bigframes-executions.png" title="Executions"/>

The 'Results' and 'Source Notebook' will also be written into the specified bucket:
<Lightbox src="/img/guides/gcp-guides/gcp-bigframes-source.png" title="Source Notebook"/>

> ## FAQs

Reviewer 2: Some of this should just go into the BigQuery config page (which we can direct users to  at the beginning of the quickstart).  This way you aren't maintaining time specific context in a quickstart. 

A quickstart should just be a tutorial on how to get set up and use it, not a blog post about the feature  or the documentation site. Please remove this entire FAQ section from the quickstart.
